accepted/system-range/system-range.md

```diff
@@ -0,0 +1,170 @@
+```
+
+### Interplay with C# syntax
+
+The syntax for ranges isn't finalized and is actively worked on. While this is
+part of a separate design group it makes sense to look at some samples to see
+how this BCL type will play with it:
+
+```
+var fiveToTen = 5..10; // Equivalent to Range.Between(5, 10)
+var fiveToRest = 5..;  // Equivalent to Range.From(5)
+var startToTen = ..10; // Equivalent to Range.To(10)
+var everything = ..;   // Equivalent to Range.All
+
+```
+
+### Using ranges in APIs
+
+Ranges provide a concise way to provide an `index` and a `length`. This allows
+for making common operations more compact and less confusing at the call side.
+For example, consider this version of the indexer for `String`:
+
+```csharp
+partial class String
+{
+    public string this[Range range] => Substring(range.Index, range.Length);
+}
+```
+
+This allows creating substrings simply by using the indexer:
+
+```csharp
+var helloWorld = "Hello, World!";
+var hello = helloWorld[..5];
+var world = helloWorld[7..5];
+```
+
+While this might appear to just be minor syntactic sugar it allows for creating
+powerful APIs, such as the indexer for `Tensor<T>` (which you can think of as
+multidimensional array where the dimensions aren't statically known).
+
+```csharp
+partial class Tenstor<T>
+{
+    public Tenstor<T> this[params Range[] ranges] { get; }
+}
+```
+
+The indexer takes a list of ranges that can be used to get slices of the tensor.
+Imagine you have a two dimensional tensor (in other words a table). Imagine it
+represents test data where each each row is a test case and the columns are
+representing the inputs with the last column containing the expected result. Now
+let's say you want to split this data into two tensors, one representing the
+inputs and one the outputs.
+
+This might look like this:
+
+```csharp
+Tensor<float> testDataTable = GetTestData();
+int columnCount = testDataTable.GetLength(1);
+int outputColumn = columnCount - 1;
+int lastDataColumn = outputColumn - 1;
+
+Tensor<float> inputTable = testDataTable[.., ..lastDataColumn];
+Tensor<float> outputTable = testDataTable[.., outputColumn..1];
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting negative ranges or element types
+  other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+
+## Design
+
+### Proposed API
+
+```csharp
+namespace System {
+    public struct Range : IEnumerable, IEnumerable<int> {
+        public static readonly Range Empty;
+        public static readonly Range All;
```

Comment:
Static properties might be better. It's not like we need to make sure there's a single instance of this struct and as a property this would be nicely inlineable, with constant propagation likely kicking in and we would avoid type initializer checks as well.

Reply:
Fair enough. I don't have a preference and for primitive types I can go either way. It's also not like we'll ever need to run code there.

---
In reply to: [163201749](https://github.com/dotnet/designs/pull/30#discussion_r163201749) [](ancestors = 163201749)